File: services/payments/routing.py
from fastapi import HTTPException

from schemas.payments_schema import PaymentProvider


def resolve_provider(country_code: str, requested_provider: PaymentProvider | None) -> PaymentProvider:
    normalized = country_code.upper()

    if normalized == "NG":
        if requested_provider is None:
            raise HTTPException(status_code=400, detail="provider is required for NG checkouts")
        if requested_provider not in (PaymentProvider.flutterwave, PaymentProvider.paystack):
            raise HTTPException(status_code=400, detail="NG supports only flutterwave or paystack")
        return requested_provider

    if requested_provider in (PaymentProvider.flutterwave, PaymentProvider.paystack):
        raise HTTPException(status_code=400, detail="Non-NG checkouts support only stripe")

    return PaymentProvider.stripe


def resolve_currency(country_code: str) -> str:
    return "NGN" if country_code.upper() == "NG" else "USD"

File: mei-next/lib/payments/routing.ts
/**
 * Provider routing by countryCode — port of `services/payments/routing.py`
 * (payments.md §1).
 * - NG → provider required, must be flutterwave or paystack (else 400).
 * - Any other country → forced stripe; explicit flutterwave/paystack → 400.
 * - Currency: NGN for NG, USD otherwise (no FX table).
 */
import { HttpError } from "@/lib/http/errors";
import type { PaymentProvider } from "./contracts";

export function resolveProvider(
  countryCode: string,
  requestedProvider?: PaymentProvider | null,
): PaymentProvider {
  const normalized = countryCode.toUpperCase();

  if (normalized === "NG") {
    if (requestedProvider === null || requestedProvider === undefined) {
      throw new HttpError(400, "provider is required for NG checkouts");
    }
    if (requestedProvider !== "flutterwave" && requestedProvider !== "paystack") {
      throw new HttpError(400, "NG supports only flutterwave or paystack");
    }
    return requestedProvider;
  }

  if (requestedProvider === "flutterwave" || requestedProvider === "paystack") {
    throw new HttpError(400, "Non-NG checkouts support only stripe");
  }

  return "stripe";
}

export function resolveCurrency(countryCode: string): string {
  return countryCode.toUpperCase() === "NG" ? "NGN" : "USD";
}
